docs/docs/examples.md

```diff
@@ -0,0 +1,317 @@
+
+Asn1.Sequence {
+    +Asn1.Int(1)
+    +OctetStringEncapsulating {
+        +(Bool(false) withImplicitTag TAG_TIME_RELATIVE)
+        +Asn1Time(Clock.System.now())
+    }
+} withImplicitTag (1337uL withClass TagClass.APPLICATION) 
+//                ↑ in reality this would be a constant ↑ 
+```
+
+The HEX-equivalent of this structure (which can be obtained by calling `.toDerHexString()`) is
+[7F8A391802010104135F2A0100170D3234303933303138313135395A](https://lapo.it/asn1js/#f4o5GAIBAQQTXyoBABcNMjQwOTMwMTgxMTU5Wg).
+
+### Parsing and Validating Tags
+
+Basic parsing is straight-forward: You have DER-encoded bytes, and feed them into `AsnElement.parse()`.
+In this example, you then examine the first child to get the number of times the operation was carried out,
+decode the first child of the OCTET STRING that follows to see if n UTC time or an int follows.
+
+Usually, though (and especially when using implicit tags), you really want to verify those tags too.
+Hence, parsing and properly validating is a bit more elaborate:
+
+```kotlin linenums="1"
+Asn1Element.parse(customSequence.derEncoded).asStructure().let { root -> 
+
+  //↓↓↓ In reality, this would be a global constant; the same as in the previous snippet ↓↓↓
+  val rootTag = Asn1Element.Tag(1337uL, tagClass = TagClass.APPLICATION, constructed = true)
+  root.assertTag(rootTag) //throws on tag mismatch
+
+  val numberOfOps = root.nextChild().asPrimitive().decodeToUInt()
+  root.nextChild().asEncapsulatingOctetString().let { timestamp ->
+    val isRelative = timestamp.nextChild().asPrimitive()
+      .decode(TAG_TIME_RELATIVE) { Boolean.decodeFromAsn1ContentBytes(it) }
+
+    val time = if (isRelative) timestamp.nextChild().asPrimitive().decodeToUInt()
+    else timestamp.nextChild().asPrimitive().decodeToInstant()
+
+    if (timestamp.hasMoreChildren() || root.hasMoreChildren())
+      throw Asn1StructuralException("Superfluous Content")
+      
+    // Everything is parsed and validated
+    TODO("Create domain object from $numberOfOps, $isRelative, and $time")
+    }
+}
+```
+
+The above snippet performs the following validations:
+
+1. Line 5 asserts the tag of the root structure
+2. Line 7 ensures that the first child is an ASN.1 primitive tagged as INT, containing an unsigned integer
+3. Line 8 execution successfully guarantees that the second child is indeed an ASN.1 OCTET STRING encapsulating
+another ASN.1 structure.
+4. Lines 9-10 verify that the first child contained in the ASN.1 OCTET STRING
+    * is an ASN.1 primitive
+    * tagged with `TAG_TIME_RELATIVE`
+    * containing an ASN.1 boolean
+5. Line 12 ensures that the next child is an ASN.1 primitive, encoding an unsigned integer (in case an `UInt` is expected)
+6. Line 13 tackles the alternative and ensures that the next child contains a properly encoded ASN.1 time
+7. Lastly, lines 15-16 make sure no additional content is present, thus fully verifying the structure as a whole.
```

Comment:
should we not do this automatically in `asEncapsulatingOctetString` after executing the user-supplied block? maybe adding the option to disable it with `strict = false`?

Reply:
?

Reply:
i misunderstood what the code does. either way, proposing:
```kotlin
fun <T> @ICouldBeWrongAboutWhatClassThisIsDefinedOn Asn1Primitive.parseEncapsulatedOctetString(block: (@NotSureWhichClass Asn1Something)->T) : T {
  val os = asEncapsulatingOctetString()
  return block(os).also {
    if (os.hasMoreChildren()) throw Asn1StructuralException("Superfluous Content")
  }
}
```

Reply:
and the same thing for the use case with `Asn1Element.parse(bla).asStructure().let { ... }`, we should have a "parse using this block and then verify that we consumed everything" helper

Reply:
now it makes sense! I agree: #184